### PR TITLE
New module - digital_ocean_project

### LINKED ISSUE #49038: Support for Digital Ocean projects
##### SUMMARY
Digital Ocean announced a new API for projects (https://developers.digitalocean.com/documentation/changelog/api-v2/introducing-the-projects-api-beta/). It is currently in beta, but it would be nice to have an Ansible module to support provisioning of projects and project resources.

##### ISSUE TYPE
- Feature request

##### COMPONENT NAME
digital_ocean_projects

### PR DESCRIPTION
**Pull request by kbreit**
##### SUMMARY
- Creates or modifies DigitalOcean projects
- API doesn't allow for project deletion, which is annoying
- Not currently idempotent
- A facts module will be created for the query

Fixes #49038

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
digital_ocean_project

##### ADDITIONAL INFORMATION
Example task
```paste below
    - name: Create new project
      digital_ocean_project:
        oauth_token: '{{do_key}}'
        state: present
        name: AnsibleTest
        description: Something or another
        purpose: IoT
        environment: Development
```

### CODE REVIEW

**Comment by kbreit** [comment]
@BondAnthony Please take a look at this module as it develops and provide feedback on it. It is my first DigitalOcean module so I'm getting used to the styles.

**Comment by kbreit** [comment]
@BondAnthony I'm close to being completed with this module and the facts one will come shortly as well. What idempotency methods do you use?

File: test/legacy/roles/test_digital_ocean/tasks/main.yml [hunk 1]
```diff
@@ -2,3 +2,5 @@
   include: tasks/sshkey.yml
 - name: Floating IPs module tests
   include: tasks/floating_ip.yml
+- name: Project tasks
+  include: tasks/project.yml
```
Comment (gundalow) [inline test/legacy/roles/test_digital_ocean/tasks/main.yml]: Missing newline
```suggestion
  include: tasks/project.yml
```

**Review by gundalow — COMMENTED** [review]
Once this is merged could you please raise a single PR to move all of `test/legacy/roles/*digital_ocean*` to `test/integration/targets/`

You will need to add an `unsupported` alias file https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/integration-aliases.html

**Comment by kbreit** [comment]
@gundalow Yes, when this is merged I'll create a new PR to move to the new integration framework. I just created issue #49505 for this effort so it's in my todo list.

Update: Initial branch for this is in my new https://github.com/kbreit/ansible/tree/digital_ocean/integration_tests branch

**Comment by kbreit** [comment]
@gundalow Are you able to kick off this CI test again? It looks like it timed out.

**Comment by kbreit** [comment]
@gundalow @BondAnthony Can I get a review on this as I'd like if it was in the 2.8 release. Thank you.

**Comment by kbreit** [comment]
@BondAnthony I'm now seeing this module doesn't correctly report `changed`. If I modify a project twice, it says it's changed, even if it hasn't. How do you handle this in other DO modules?

**Comment by kbreit** [comment]
Since this has been sitting a while and it doesn't seem like Digital Ocean modules are in development, I'm going to close this for now.

**Comment by winem** [comment]
> @BondAnthony I'm now seeing this module doesn't correctly report `changed`. If I modify a project twice, it says it's changed, even if it hasn't. How do you handle this in other DO modules?

Hi @kbreit , I saw that you closed the PR but I have a bit too much time at the moment and wanted to contribute to Ansible again... and the DO plugins seem to be a perfect candidate for it. 

I did some research and I think this is an "issue" with the DO-API and a proper fix is actually quite tricky. 

Observation 1: 
The PUT method always returns 200. 200 OK and 204 No Content are valid responses for PUT requests that update a resource.So we can not use the RC to distinguish whether the resource (= project) was updated or not.

Observation 2:
A PUT to update the project updates the project on the first try but the returned timestamp for `updated_at` is not modified. You'll just see the correct `updated_at` timestamp on further calls. 

Observation 3:
Unfortunately it's the same behaviour as described in Observation 2 above for PATCH request.

So the only solution (which has a big drawback; see below) that comes to my mind to fix the "always changed issue" would be a 2nd request after the PUT to see if the timestamp was updated and the timestamp is within the last n seconds. But this is unreliable as you can not really say if the update was done due to request during the playbook run or if it was done due to any other change.

I created a ticket at DO and asked them for some input. Maybe we have to post it to ideas.diitalocean.com and hope that there will be a way to determine whether a resource was updated or not in the future. :)

**Pull request closed without merging**